scipy/linalg/_matfuncs_sqrtm.c

```diff
@@ -0,0 +1,1009 @@
+    // Workspace computation and allocation
+    // --------------------------------------------------------------------
+    int info = 0, sdim = 0, lwork = -1, intn = (int)n;
+    float tmp_float = 0.0f, one = 1.0f, zero = 0.0f;
+    sgees_("V", "N", NULL, &intn, NULL, &intn, &sdim, NULL, NULL, NULL, &intn, &tmp_float, &lwork, NULL, &info);
+    // Improbable to fail at lwork query but check anyway
+    if (info != 0) { *sq_info = -100; return;}
+    lwork = (int)tmp_float;
+    // 2n*n + 2n*n for data and vs for potentially converting to complex if needed
+    // n + n for wr, wi (needed for gees calls)
+    // lwork for work
+    size_t buffer_size = 4*n*n + 2*n + lwork;
+    float* restrict buffer = malloc(buffer_size*sizeof(float));
+    if (buffer == NULL) { *sq_info = -101; return; }
+
+    // --------------------------------------------------------------------
+    // Pointers to the variables inside the buffer for easy access
+    // --------------------------------------------------------------------
+
+    // Since we already allocated n*2n for the input, we can use the first and
+    // second n*n for C to F conversion when needed.
+    float* restrict data = &buffer[0];
+    float* restrict data2 = &buffer[n*n];
+    float* restrict vs = &buffer[2*n*n];
+    float* restrict wr = &buffer[4*n*n];
+    float* restrict wi = &buffer[4*n*n + n];
+    float* restrict work = &buffer[4*n*n + 2*n];
+    // Cover the first 2n*n and vs area with a complex pointer just in case
+    SCIPY_C* complex_data = &((SCIPY_C*)buffer)[0];
+    SCIPY_C* complex_vs = &((SCIPY_C*)buffer)[n*n];
+
+
+    /*====================================================================
+    |                    MAIN nxn SLICE LOOP                             |
+    ====================================================================*/
+    for (npy_intp idx = 0; idx < outer_size; idx++) {
+        /*------------------------------------------------------------------
+        Copy the current slice into buffer in F-layout for LAPACK calls. Note
+        that the input can be in C or F layout and not necessarily contiguous.
+        Since the input is predominantly C contiguous, we copy the data in
+        C-layout first and then transpose it to F-layout.
+
+        The output is always a C contiguous array hence we visit slices in
+        C-order for easy insertion to the output array; by just offsetting
+        idx*n*n * sizeof(dtype). For this reason, we  start the offset
+        calculation from the inner-most outer dimension.
+        ------------------------------------------------------------------*/
+        npy_intp offset = 0;
+        npy_intp temp_idx = idx;
+        for (int i = ndim - 3; i >= 0; i--) {
+            offset += (temp_idx % shape[i]) * strides[i];
+            temp_idx /= shape[i];
+        }
+        float* restrict slice_ptr = (float*)(Am_data + (offset/sizeof(float)));
+
+        for (int i = 0; i < n; i++) {
```

Comment:
On re-reading this to actually make sure I understand the iteration over slices: internal loop variable `i` is used twice:  line 89 above and this line 95. Even if it is actually legal in C (is it?), consider not reusing the same name for the loop variable in nested loops.

This is of course minor, can be pushed to a follow-up. Would be great to have this PR in regardless.

Reply:
This is something I used to wonder earlier and essentially you are asking what is the difference between 

```C
int i;
for (i = 0; ...) { ...}
```

and 

```C
for (int i = 0; ...) { ... }
```

In the first one, `i` lives in the scope of the definition after the loop, so you can use again without defining, the second one has `i` only defined in the loop and is not defined outside that particular loop. You can try 

```C
#include <stdio.h>

int main() {
    for (int i = 0; i < 5; i++) { printf("i = %d\n", i); }
    printf("i = %d\n", i);
    return 0;
}
```

to get a compiler error. Is this a clarification or am I just writing past you?

Reply:
A bit past, sadly.

Your example

```
int i;
for(i=0, ....
```

shows that it is probably legal to write

```
for (int i=0; i < 10; i++) {
   for(int i=0; i < 10; i++){
     printf("%d\n", i);
   }
}
```

OK, it's something we _can_ do; is this something we should do though?

Reply:
Ok, rereading this again. Lines 89 and 93 are not nested. 
The loops are one after the other.
Case closed.
Sorry for the noise.